suite/tests/client-interface/stolen-reg-index.appdll.c

```diff
@@ -0,0 +1,96 @@
+/* **********************************************************
+ * Copyright (c) 2021 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* App binary for stolen-reg-index test. Executes two indexed memory
+ * instructions, one using X28 for index register, the other W28. Intention is
+ * for DR to call drutil_insert_get_mem_addr() in order to test the
+ * 'if (index == stolen)' clause in drutil_insert_get_mem_addr_arm() in the
+ * case of W28.
+ */
+
+/***************************************************************************/
+#ifndef ASM_CODE_ONLY /* C code */
+
+#    include "tools.h"
+
+/* In asm code. */
+void
+indexed_mem_test(int *val);
+
+void
+run_tests(void)
+{
+    int value = 41;
+    indexed_mem_test(&value);
+    if (value != 42) {
+        fprintf(stderr, "indexed_mem_test() failed with %d, expected 42.\n", value);
+        fflush(stderr);
+    } else {
+        fprintf(stdout, "indexed_mem_test() passed.\n");
+        fflush(stdout);
+    }
+}
+
+int __attribute__((constructor)) so_init(void)
+{
+    run_tests();
+    return 0;
+}
+
+#else /* asm code *************************************************************/
+#    include "asm_defines.asm"
+/* clang-format off */
+START_FILE
+#define FUNCNAME indexed_mem_test
```

Comment:
I was going to say that with just one asm file and without needing Windows support, this could be an inline asm function added to drutil-test.c to leverage whatever address verification is there too, rather than a separate file.  But: it looks like today's drutil-test.c is drmgr-test.c.  Huh.  That's only to get the rep str asm code I assume: it doesn't need the other stuff.  Maybe a new drutil-test.c should be made that has the rep str for x86 and stolen reg for arm.